doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -0,0 +1,970 @@
+
+.. productionlist:: coq
+   ltac2_term             : match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
+                    : lazy_match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
+                    : multi_match! [ reverse ] goal with `goalmatching` ... `goalmatching` end
+   goalmatching     : | [ `hypmatching` ... `hypmatching` |- `constrpattern` ] => `ltac2_term`
+   hypmatching      : `lident` : `constrpattern`
+
+Variables from :n:`@hypmatching` and :n:`@constrpattern` are bound in the body of the
+branch. Their types are:
+
+- ``constr`` for pattern variables appearing in a :n:`@constr`
+- ``Pattern.context`` for variables binding a context
+- ``ident`` for variables binding a hypothesis name.
+
+The same identifier caveat as in the case of matching over constr applies, and
+this features has the same semantics as in Ltac1. In particular, a ``reverse``
+flag can be specified to match hypotheses from the more recently introduced to
+the least recently introduced one.
+
+Notations
+---------
+
+Notations are the crux of the usability of Ltac1. We should be able to recover
+a feeling similar to the old implementation by using and abusing notations.
+
+Scopes
+~~~~~~
+
+A scope is a name given to a grammar entry used to produce some Ltac2 expression
+at parsing time. Scopes are described using a form of S-expression.
+
+.. prodn::
+   scope ::= @string %| @int %| @lident ({+, @scope})
+
+A few scopes contain antiquotation features. For sake of uniformity, all
+antiquotations are introduced by the syntax :n:`$@var`.
+
+The following scopes are built-in.
+
+- :n:`constr`:
+
+  + parses :n:`c = @constr` and produces :n:`constr:(c)`
+
+- :n:`ident`:
+
+  + parses :n:`id = @ident` and produces :n:`ident:(id)`
+  + parses :n:`$(x = @ident)` and produces the variable :n:`x`
+
+- :n:`list0(@scope)`:
+
+  + if :n:`@scope` parses :n:`@entry`, parses :n:`(x₀, ..., xₙ = {* @entry})` and produces
+    :n:`[x₀; ...; xₙ]`.
+
+- :n:`list0(@scope, sep = @string)`:
+
+  + if :n:`@scope` parses :n:`@entry`, parses `(x₀ = @entry, sep, ..., sep, xₙ = @entry)`
+    and produces `[x₀; ...; xₙ]`.
+
+- :n:`list1`: same as :n:`list0` (with or without separator) but parses :n:`{+ @entry}` instead
+  of `ENTRY*`.
+
+- :n:`opt(@scope)`
+
+  + if :n:`@scope` parses :n:`@entry`, parses :n:`{? @entry}` and produces either :n:`None` or
+    :n:`Some x` where :n:`x` is the parsed expression.
+
+- :n:`self`:
+
+  + parses a Ltac2 expression at the current level and return it as is.
+
+- :n:`next`:
+
+  + parses a Ltac2 expression at the next level and return it as is.
+
+- :n:`tactic(n = @int)`:
+
+  + parses a Ltac2 expression at the provided level :n:`n` and return it as is.
+
+- :n:`thunk(@scope)`:
+
+  + parses the same as :n:`scope`, and if :n:`e` is the parsed expression, returns
+    :n:`fun () => e`.
+
+- :n:`STRING`:
+
+  + parses the corresponding string as a CAMLP5 IDENT and returns :n:`()`.
```

Comment:
What does this mean?

Reply:
Not sure I understand the question, but e.g. writing "foo" will parse the identifier `foo`, and the result will be bound to the unique value of type unit.

Reply:
I meant: do you have to talk about `CAMLP5 IDENT` here?

Reply:
Not necessarily, but the point of the comment is that this shares the same grammar as Coq idents. Should I just write "an identifier"?